File: libgomp/testsuite/libgomp.fortran/examples-4/simd-2.f90
! { dg-do run { target vect_simd_clones } }
! { dg-additional-options "-msse2" { target sse2_runtime } }
! { dg-additional-options "-mavx" { target avx_runtime } }

module SIMD2_mod
contains
  function add1(a,b,fact) result(c)
  !$omp declare simd(add1) uniform(fact)
     double precision :: a,b,fact, c
     c = a + b + fact
  end function

  function add2(a,b,i, fact) result(c)
  !$omp declare simd(add2) uniform(a,b,fact) linear(i:1)
     integer,          value        :: i
     double precision, dimension(:) :: a, b
     double precision               :: fact, c
     c = a(i) + b(i) + fact
  end function

  subroutine work(a, b, n )
     implicit none
     double precision           :: a(n),b(n), tmp
     integer                    :: n, i

     !$omp simd private(tmp)
     do i = 1,n
        tmp  = add1(a(i), b(i), 1.0d0)
        a(i) = add2(a,    b, i, 1.0d0) + tmp
        a(i) = a(i) + b(i) + 1.0d0
     end do
  end subroutine

  subroutine work_ref(a, b, n )
     implicit none
     double precision           :: a(n),b(n), tmp
     integer                    :: n, i

     do i = 1,n
        tmp  = add1(a(i), b(i), 1.0d0)
        a(i) = add2(a,    b, i, 1.0d0) + tmp
        a(i) = a(i) + b(i) + 1.0d0
     end do
  end subroutine

  subroutine check (a, b, n)
      integer :: i, n
      double precision, parameter :: EPS = 0.0000000000001
      double precision :: diff, a(*), b(*)
      do i = 1, n
        diff = a(i) - b(i)
        if (diff > EPS .or. -diff > EPS) call abort
      end do
  end subroutine
end module

program main
   use SIMD2_mod
   integer, parameter :: N=32
   integer :: i
   double precision   :: a(N), b(N), a_ref(N)
   do i = 1,N
      a(i) = i-1
      a_ref(i) = a(i)
      b(i) = N-(i-1)
   end do

   call work(a, b, N )
   call work_ref(a_ref, b, N )

   call check(a, a_ref, N )
end program

File: libgomp/testsuite/libgomp.c/examples-4/simd-2.c
/* { dg-do run { target vect_simd_clones } } */
/* { dg-additional-options "-msse2" { target sse2_runtime } } */
/* { dg-additional-options "-mavx" { target avx_runtime } } */

#define N 100
#define EPS 0.0000000000000001

#include <stdlib.h>

void init(double *a, double *a_ref, double *b, int n)
{
   int i;
   for ( i=0; i<N; i++ )
   {
      a[i] = i;
      a_ref[i] = i;
      b[i] = N-i;
   }
}

#pragma omp declare simd uniform(fact)
double add1(double a, double b, double fact)
{
   double c;
   c = a + b + fact;
   return c;
}

#pragma omp declare simd uniform(a,b,fact) linear(i:1)
double add2(double *a, double *b, int i, double fact)
{
   double c;
   c = a[i] + b[i] + fact;
   return c;
}

#pragma omp declare simd uniform(fact) linear(a,b:1)
double add3(double *a, double *b, double fact)
{
   double c;
   c = *a + *b + fact;
   return c;
}

void work( double *a, double *b, int n )
{
   int i;
   double tmp;
   #pragma omp simd private(tmp)
   for ( i = 0; i < n; i++ ) {
      tmp  = add1( a[i],  b[i], 1.0);
      a[i] = add2( a,     b, i, 1.0) + tmp;
      a[i] = add3(&a[i], &b[i], 1.0);
   }
}

void work_ref( double *a, double *b, int n )
{
   int i;
   double tmp;
   for ( i = 0; i < n; i++ ) {
      tmp  = add1( a[i],  b[i], 1.0);
      a[i] = add2( a,     b, i, 1.0) + tmp;
      a[i] = add3(&a[i], &b[i], 1.0);
   }
}

void check (double *a, double *b)
{
  int i;
  for (i = 0; i < N; i++)
    if (a[i] - b[i] > EPS || b[i] - a[i] > EPS)
      abort ();
}


int main ()
{
   int i;
   double a[N], a_ref[N], b[N];

   init(a, a_ref, b, N);

   work(a, b, N );
   work_ref(a_ref, b, N );

   check(a, a_ref);

   return 0;
}
